File: backend/app/schemas/character.py
from pydantic import BaseModel
from uuid import UUID
from datetime import datetime
from typing import Optional, Any


class CharacterCreate(BaseModel):
    project_id: UUID
    name: str
    role: Optional[str] = None
    gender: Optional[str] = None
    estimated_age: Optional[str] = None
    physical_description: Optional[str] = None


class CharacterResponse(BaseModel):
    id: UUID
    project_id: UUID
    name: str
    role: Optional[str] = None
    gender: Optional[str] = None
    estimated_age: Optional[str] = None
    physical_description: Optional[str] = None
    personality_summary: Optional[str] = None
    mbti: Optional[str] = None
    mbti_confidence: Optional[int] = None
    speech_pattern: Optional[str] = None
    emotional_arc: Optional[dict[str, Any]] = None
    reference_image_url: Optional[str] = None
    visual_description: Optional[str] = None
    video_prompt_fragment: Optional[str] = None
    face_keywords: Optional[list[str]] = None
    # "ref_rejected" = the uploaded photo was refused by the image service's
    # content filter (recognizable public figure) — warn before casting money
    plate_status: Optional[str] = None
    # non-human cast member (computed by the router from the descriptions, not
    # a DB column): the Characters page splits humans from creatures at FIRST
    # paint with this — waiting for the casting bible made the pet visibly
    # jump sections when the bible query landed later
    creature: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/lib/types/character.ts
export interface Character {
  id: string;
  project_id: string;
  name: string;
  role: "PROTAGONIST" | "ANTAGONIST" | "SUPPORTING" | "MINOR" | null;
  gender: string | null;
  estimated_age: string | null;
  physical_description: string | null;
  personality_summary: string | null;
  mbti: string | null;
  mbti_confidence: number | null;
  speech_pattern: string | null;
  emotional_arc: EmotionalArc | null;
  reference_image_url: string | null;
  visual_description: string | null;
  video_prompt_fragment: string | null;
  face_keywords: string[] | null;
  /** "ref_rejected" = the uploaded photo was refused by the image service's
   *  content filter (recognizable public figures are blocked) */
  plate_status: string | null;
  /** non-human cast member, computed server-side; rides the characters list
   *  so the humans/animals split is right on FIRST paint (the casting bible
   *  resolves later and used to make the pet visibly jump sections) */
  creature?: boolean;
  created_at: string;
}

export interface EmotionalArc {
  start: string;
  midpoint: string;
  end: string;
}

export interface CharacterCreate {
  project_id: string;
  name: string;
  role?: string;
  gender?: string;
  estimated_age?: string;
  physical_description?: string;
}
